review-problems-sol.ml

```diff
@@ -179,6 +179,13 @@ let rec pairs l = match l with
   |Cons(x,Nil)-> Nil
   |Cons(x,Cons(y,xs))-> Cons((x,y),pairs Cons(y,xs))
 
+let rec pairs_k ls return = 
+  match ls with
+  |Nil -> return Nil
+  |Cons (a, Nil) -> return Nil
+  |Cons (a,Cons(b,ls')) -> pairs_k ls' (fun x -> return (Cons((a,b),x)))
```

Comment:
this solution is incorrect. If we run it on `[1;2;3;4;5]`, we get
```ocaml
[(1,2),(3,4)]
```
but we should get
```ocaml
[(1,2),(2,3),(3,4),(4,5)]
```
To get the right solution, look closely at the non-CPS solution for `pairs_k` and see what list the recursive call is made on.

Reply:
Oh no, I must have messed it up when refactoring. I will correct this in just a bit!